WordPress/wordpress-playground
# Browser: php.mount() support for OPFS/Native FS handles

## Context of this PR

It enables mounting OPFS and local directory handles in the browser version of Playground using the same abstraction as we use for mounting local directories in Node.js. This, in turn, enables using OPFS mounts in the [boot protocol](https://github.com/WordPress/wordpress-playground/pull/1390) and relying on the same general code paths and abstractions.

## What's the technical change?

Ships a `DirectoryHandleMount` class that can be used to mount OPFS using the `php.mount()` method:

```ts
await php.mount(
	php.documentRoot,
	createDirectoryHandleMountHandler(opfsHandle);
);
```

## Implementation details

DirectoryHandleMount is not a "real" mount in that it doesn't actually plug in an Emscripten FS implementation into the PHP Emscripten module. Instead, it copies all the OPFS files into Playground MEMFS (or the other way around). After that initial sync, it journals all the MEMFS filesystem operations and replays them in OPFS. This is good enough for the in-browser directory handle, since the underlying files aren't going to change on their own, but for the Local Filesystem directory handle we also have an explicit "Synchronize files" button to bring any local changes back into Playground.

Once Emscripten supports asynchronous filesystem operations via wasmfs, we'll be about to get rid of the journal and mount an OPFS implementation to directly delegate writes, reads, etc to the underlying filesystem.

### Other changes

This PR removes the `Mountable` interface in favor of functional API to simplify the implementation and remove any state management concerns.

**Before:**

```ts
php.mount(dir, new NodeFSMount(dir));
```

**After:**

```ts
php.mount(dir, createNodeFsMountHandler(dir));
```

## Testing Instructions

Open local Playground and change the storage option to "browser", make some visible changes, refresh Playground a few times, confirm the site still loads and the changes are still around. Close the browser, reopen it, confirm it still works.

Them repeat the test with the local directory storage option.

Also, confirm the following Playground CLI command works and sets up a site in a local `new-wordpress-site` directory:

```bash
bun packages/playground/cli/src/cli.ts server --mount-before-install=`pwd`/new-wordpress-site:/wordpress
```
